File: tools/ai_tells.py
#!/usr/bin/env python3
"""
ai_tells.py — find the AI writing patterns in the site's own words.

The patterns and the words to watch come from the humanizer skill in
.claude/skills/humanizer, which is itself built from Wikipedia's "Signs
of AI writing" (WikiProject AI Cleanup).

This only finds candidates. It cannot tell a real tell from a false
positive — the skill's own list of what not to flag is half its length,
and a word is only evidence when it sits with others. So the output is a
reading list, not a verdict.

    python3 tools/ai_tells.py                 the articles
    python3 tools/ai_tells.py --all           the site copy too
"""
from __future__ import annotations

import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent

# §1 inflated importance, §3 shallow -ing analysis, §4 sales language,
# §5 vague sources, §7 stock words, §8 avoiding is/are, §27 deeper truth,
# §28 announcing, §32 formulaic sayings, §33 fake candour
PATTERNS = [
    ("inflated importance", r"\b(stands as|serves as|is a testament|a vital|crucial|pivotal|underscor\w+|highlight(?:s|ing) (?:its|the) (?:importance|significance)|evolving landscape|indelible mark|deeply rooted|marking a|represents a shift|key turning point)\b"),
    ("shallow -ing analysis", r",\s+(highlighting|underscoring|emphasi[sz]ing|ensuring|reflecting|symboli[sz]ing|contributing to|fostering|showcasing|encompassing)\b"),
    ("sales language", r"\b(boasts a|vibrant|profound|nestled|in the heart of|groundbreaking|renowned|breathtaking|must-visit|stunning|commitment to|exemplifies)\b"),
    ("vague source", r"\b(industry reports|observers have|experts (?:argue|believe|say)|some critics|several sources|studies show)\b"),
    ("stock AI word", r"\b(delve|intricate|intricacies|interplay|tapestry|testament|leverage|robust|seamless|holistic|myriad|plethora|garner|utilize|facilitate|paradigm|synerg\w+)\b"),
    ("avoiding is/are", r"\b(serves as a|stands as a|boasts|features a|offers a range)\b"),
    ("not X but Y", r"\b(not (?:just|only|merely) [^.,;]{2,40}(?:,|;| but)\s*(?:it|they|this)?'?s?\s*(?:also )?)\b"),
    ("deeper truth", r"\b(the real question is|at its core|in reality|what really matters|fundamentally,|the deeper issue|the heart of the matter)\b"),
    ("announcing the point", r"\b(let'?s (?:dive|explore|break this down|look at)|here'?s what you need to know|without further ado|quick note)\b"),
    ("formulaic saying", r"\b(is the (?:language|currency|architecture) of|becomes a trap)\b"),
    ("fake candour", r"(^|[.!?]\s+)(Honestly[?,]|Look,|Here'?s the thing|The thing is,|Let'?s be honest|Real talk)"),
    ("chatbot artifact", r"\b(I hope this helps|Of course!|Certainly!|You'?re absolutely right|let me know if|Would you like me to)\b"),
    ("knowledge-limit hedge", r"\b(as of my|up to my last|while specific details|based on available information|it is believed that|likely (?:grew up|studied|began))\b"),
    ("too many qualifiers", r"\b(could potentially|might arguably|it'?s also possible that|in some cases it may)\b"),
    ("generic positive ending", r"\b(the future looks bright|exciting times|a step in the right direction|journey toward)\b"),
    ("answering nobody", r"\b(don'?t get me wrong|to be clear,|this is not to say|I'?m not (?:saying|arguing)|some might say)\b"),
    ("fake alternative", r"\b(a tempting (?:option|approach)|one might be tempted|an obvious approach would be|it would be easy to just)\b"),
    ("em or en dash", r"[—–]"),
    ("curly quote", r"[“”‘’]"),
    ("emoji heading", r"^#{1,6}\s*[\U0001F300-\U0001FAFF]"),
    # §17: every main word capitalised. Written as "three or more
    # capitalised words in a row", which is what title case looks like and
    # what an ordinary sentence-case heading never does. The first attempt
    # at this had no word boundaries and matched every heading in the
    # journal, which is the reverse of useful.
    ("title case heading", r"^#{2,6}\s+\b[A-Z][a-z]+\b(?:\s+\b[A-Z][a-z]+\b){2,}\s*$"),
    ("bold mini-heading list", r"^\s*[-*]\s+\*\*[^*]+:\*\*"),
]

# Case matters for exactly two of these, and they are the two about
# capitalisation. Compiling the whole list with re.I made [A-Z] match
# lowercase, so the title-case rule matched every sentence-case heading
# in the journal — 40 of the 47 "candidates" in its first run.
CASE_SENSITIVE = {"title case heading", "fake candour"}

COMPILED = [
    (name, re.compile(rx, re.M if name in CASE_SENSITIVE else re.I | re.M))
    for name, rx in PATTERNS
]


def scan(path: Path) -> list[tuple[str, int, str]]:
    text = path.read_text(encoding="utf-8")
    # front matter is metadata, not prose
    text = re.sub(r"^---\n.*?\n---\n", "", text, flags=re.S)
    hits = []
    for name, rx in COMPILED:
        for m in rx.finditer(text):
            line = text.count("\n", 0, m.start()) + 1
            start = max(0, m.start() - 50)
            before = " ".join(text[start:m.start()].split())
            after = " ".join(text[m.end():m.end() + 50].split())
            got = " ".join(m.group(0).split())
            hits.append((name, line, f"…{before} «{got}» {after}…"))
    return sorted(hits, key=lambda h: h[1])


def site_copy() -> str:
    """Every string the site puts in front of a reader, out of the seed.

    The articles are files; the rest of the site's words are rows in D1,
    baked into lib/seed.js as INSERTs. Scanning the .js as text would
    flag the code around them, so the values are pulled out first.
    """
    src = (ROOT / "lib" / "seed.js").read_text(encoding="utf-8")
    # the seeds carry their strings as SQL literals inside JS strings
    parts = re.findall(r"'((?:[^'\\]|\\.|'')+)'", src)
    words = [p.replace("''", "'").replace("\\n", "\n") for p in parts]
    # anything with a space and a letter is prose; the rest is slugs,
    # column names and enum values
    return "\n\n".join(w for w in words if " " in w and re.search(r"[a-z]{4}", w))


def main() -> int:
    files = sorted((ROOT / "content" / "articles").glob("*.md"))
    extra: list[tuple[str, str]] = []
    if "--all" in sys.argv:
        files += [ROOT / "lib" / "legal.js", ROOT / "index.html", ROOT / "book.html"]
        extra.append(("lib/seed.js (the site's own copy)", site_copy()))

    total = 0
    for f in files:
        if not f.exists():
            continue
        hits = scan(f)
        if not hits:
            continue
        total += len(hits)
        print(f"\n{f.relative_to(ROOT)}  ({len(hits)})")
        for name, line, snippet in hits:
            print(f"  {line:>4}  {name:<24} {snippet}")

    for label, text in extra:
        hits = []
        for name, rx in COMPILED:
            for m in rx.finditer(text):
                got = " ".join(m.group(0).split())
                start = max(0, m.start() - 60)
                ctx = " ".join(text[start:m.end() + 60].split())
                hits.append((name, 0, f"«{got}» in …{ctx}…"))
        if not hits:
            continue
        total += len(hits)
        print(f"\n{label}  ({len(hits)})")
        for name, _, snippet in hits:
            print(f"        {name:<24} {snippet}")

    print(f"\n{total} candidates. Read each one; most words on these lists are")
    print("ordinary English, and the skill's own advice is that a single hit")
    print("proves nothing.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/ai_tells.mjs
/**
 * ai_tells.mjs — find the signs of AI writing in the site's own words.
 *
 * The patterns live in assets/js/tells.js, which the Worker also uses to
 * refuse a draft, so what this reports and what the server enforces
 * cannot drift apart. That was the reason for moving this off Python:
 * two lists in two languages is one list plus a disagreement.
 *
 *   node tools/ai_tells.mjs            the articles
 *   node tools/ai_tells.mjs --all      the site's own copy too
 */
import { readFileSync, readdirSync } from 'node:fs';
import { judge, findTells } from '../assets/js/tells.js';

const all = process.argv.includes('--all');

/*
 * Page titles were the one piece of the site's writing nothing scanned.
 * The detector ran on articles and on the copy in seed.js, and the site
 * shipped with an em dash in the <title> of the home page, the 404 and
 * the booking page for months. A title is the most-read sentence on a
 * page — it is what a search result and a browser tab show — so it goes
 * first, and it is checked on every run rather than only under --all.
 */
const TITLED = [
  'index.html', '404.html', 'book.html', 'studio.html',
  'lib/templates.js', 'lib/plainpage.js',
];

/* \u2014 in a source file is the same character to a reader and a
   different one to a regex, which is exactly how the 404's title kept its
   em dash through the first pass of this. */
const unescape = (t) => t.replace(/\\u([0-9a-fA-F]{4})/g,
  (_, h) => String.fromCharCode(parseInt(h, 16)));

export function pageTitles() {
  const out = [];
  for (const f of TITLED) {
    let src = '';
    try { src = unescape(readFileSync(f, 'utf8')); } catch { continue; }
    for (const m of src.matchAll(/<title>([^<]+)<\/title>/g)) out.push([f, m[1]]);
    for (const m of src.matchAll(/(?:og:title|twitter:title)"\s+content="([^"]+)"/g)) out.push([f, m[1]]);
    for (const m of src.matchAll(/content="([^"]+)"\s+property="og:title"/g)) out.push([f, m[1]]);
    // the ones built in JS: title: `...`
    for (const m of src.matchAll(/title:\s*`([^`]+)`/g)) out.push([f, m[1]]);
  }
  return out;
}

const files = readdirSync('content/articles')
  .filter((f) => f.endsWith('.md'))
  .sort()
  .map((f) => [`content/articles/${f}`, readFileSync(`content/articles/${f}`, 'utf8')]);

if (all) {
  // The articles are files; the rest of the site's words are rows in D1,
  // baked into lib/seed.js as INSERTs. Scanning the .js as text would
  // flag the code around them, so the literals come out first.
  const src = readFileSync('lib/seed.js', 'utf8');
  const words = [...src.matchAll(/'((?:[^'\\]|\\.|'')+)'/g)]
    .map((m) => m[1].replaceAll("''", "'").replaceAll('\\n', '\n'))
    .filter((w) => w.includes(' ') && /[a-z]{4}/.test(w));
  files.push(['lib/seed.js (the site\'s own copy)', words.join('\n\n')]);
  for (const f of ['lib/legal.js', 'book.html']) {
    files.push([f, readFileSync(f, 'utf8')]);
  }
}

const titles = pageTitles();
const titleFaults = titles
  .map(([f, t]) => [f, t, findTells(t).filter((x) => x.weight === 'hard')])
  .filter(([, , bad]) => bad.length);

let hard = titleFaults.length, soft = 0;
if (titleFaults.length) {
  console.log('\npage titles');
  for (const [f, t, bad] of titleFaults) {
    console.log(`  ${f}\n    ${t}\n    ${bad.map((b) => b.say).join(' ')}`);
  }
} else {
  console.log(`\n${titles.length} page titles, all clean`);
}

for (const [name, text] of files) {
  const found = findTells(text);
  if (!found.length) continue;
  hard += found.filter((f) => f.weight === 'hard').length;
  soft += found.filter((f) => f.weight === 'soft').length;
  console.log(`\n${name}  (${found.length})`);
  for (const f of found) {
    console.log(`  ${String(f.line).padStart(4)}  ${f.weight === 'hard' ? 'REFUSE' : 'read  '}  ${f.id.padEnd(20)} ${f.context}`);
  }
}

console.log(`\n${hard} would be refused, ${soft} to read.`);
console.log('The second number is a reading list, not a verdict: most words on');
console.log('those lists are ordinary English, and one of anything proves nothing.');

/*
 * A title fault is the only thing here that fails the run.
 *
 * Everything else is a reading list on purpose: prose written by a person
 * trips these patterns all the time and a scanner that blocks a commit
 * over one "however" is a scanner people delete. A title is different. It
 * is short, there are sixteen of them, every one was written deliberately,
 * and it is the sentence a search result shows. There is no honest reason
 * for an em dash to be in one.
 */
process.exit(titleFaults.length ? 1 : 0);
